linera-views/src/backends/lru_caching.rs

```diff
@@ -22,25 +36,259 @@ use crate::memory::MemoryStore;
+});
+
+#[cfg(with_metrics)]
+/// The total number of cache successes
+static NUM_CACHE_VALUE_SUCCESS: LazyLock<IntCounterVec> = LazyLock::new(|| {
+    prometheus_util::register_int_counter_vec(
+        "num_cache_value_success",
+        "Number of value cache success",
+        &[],
+    )
+    .expect("Counter creation should not fail")
+});
+
+#[cfg(with_metrics)]
+/// The total number of value cache faults
+static NUM_CACHE_FIND_FAULT: LazyLock<IntCounterVec> = LazyLock::new(|| {
+    prometheus_util::register_int_counter_vec(
+        "num_cache_find_fault",
+        "Number of find cache faults",
+        &[],
+    )
+    .expect("Counter creation should not fail")
 });
 
 #[cfg(with_metrics)]
 /// The total number of cache successes
-static NUM_CACHE_SUCCESS: LazyLock<IntCounterVec> = LazyLock::new(|| {
-    prometheus_util::register_int_counter_vec("num_cache_success", "Number of cache success", &[])
-        .expect("Counter creation should not fail")
+static NUM_CACHE_FIND_SUCCESS: LazyLock<IntCounterVec> = LazyLock::new(|| {
+    prometheus_util::register_int_counter_vec(
+        "num_cache_find_success",
+        "Number of find cache success",
+        &[],
+    )
+    .expect("Counter creation should not fail")
 });
 
+#[derive(Clone, Hash, Eq, PartialEq, Ord, PartialOrd)]
+enum CacheEntry {
+    Find,
+    Value,
+}
+
+enum ValueCacheEntry {
+    DoesNotExist,
+    Exists,
+    Value(Vec<u8>),
+}
+
+impl ValueCacheEntry {
+    fn from_contains_key(test: bool) -> ValueCacheEntry {
+        match test {
+            false => ValueCacheEntry::DoesNotExist,
+            true => ValueCacheEntry::Exists,
+        }
+    }
+
+    fn from_read_value(result: &Option<Vec<u8>>) -> ValueCacheEntry {
+        match result {
+            None => ValueCacheEntry::DoesNotExist,
+            Some(vec) => ValueCacheEntry::Value(vec.to_vec()),
+        }
+    }
+
+    fn get_contains_key(&self) -> bool {
+        match self {
+            ValueCacheEntry::DoesNotExist => false,
+            ValueCacheEntry::Exists => true,
+            ValueCacheEntry::Value(_) => true,
+        }
+    }
+
+    fn get_read_value(&self) -> Option<Option<Vec<u8>>> {
+        match self {
+            ValueCacheEntry::DoesNotExist => Some(None),
+            ValueCacheEntry::Exists => None,
+            ValueCacheEntry::Value(vec) => Some(Some(vec.clone())),
+        }
+    }
+
+    fn size(&self) -> usize {
+        match self {
+            ValueCacheEntry::DoesNotExist => 0,
+            ValueCacheEntry::Exists => 0,
+            ValueCacheEntry::Value(vec) => vec.len(),
+        }
+    }
+}
+
+enum FindCacheEntry {
+    Keys(Vec<Vec<u8>>),
+    KeyValues(Vec<(Vec<u8>, Vec<u8>)>),
+}
+
+impl FindCacheEntry {
+    fn get_find_keys(&self, key_prefix: &[u8]) -> Vec<Vec<u8>> {
+        let delta = key_prefix.len();
+        let mut keys = Vec::new();
+        match self {
+            FindCacheEntry::Keys(vec) => {
+                for key in vec {
```

Comment:
You are right in finding out the lack of performance of this section of the code. It is true that the vectors are ordered.

However, it is not a good idea to use a `BTreeSet` / `BTreeMap` since the memory usage of those containers is larger than the one of vectors. Instead the use of `binary_search_by` is the mandated solution. Let me look at this.